ops/hookcmds.py

```diff
@@ -0,0 +1,995 @@
+    return SecretInfo(
+        id=id,
+        label=data.get('label'),
+        description=data.get('description'),
+        expiry=datetime.datetime.fromisoformat(data['expiry']) if data.get('expiry') else None,
+        rotation=SecretRotate(data['rotation']) if data.get('rotation') else None,
+        rotates=datetime.datetime.fromisoformat(data['rotates']) if data.get('rotates') else None,
+        revision=data['revision'],
+    )
+
+
+def secret_remove(id: str, *, revision: int | None = None):
+    """Remove an existing secret."""
+    args = ['secret-remove', id]
+    if revision is not None:
+        args.extend(['--revision', str(revision)])
+    _run(*args)
+
+
+def secret_revoke(id: str, *, relation_id: int | None, app: str | None, unit: str | None = None):
+    """Revoke access to a secret."""
+    args = ['secret-revoke']
+    if relation_id is not None:
+        args.extend(['--relation', str(relation_id)])
+    if app is not None:
+        args.extend(['--app', app])
+    if unit is not None:
+        args.extend(['--unit', unit])
+    args.append(id)
+    _run(*args)
+
+
+def secret_set(
+    id: str,
+    *,
+    content: dict[str, str] | None = None,
+    label: str | None = None,
+    description: str | None = None,
+    expire: datetime.datetime | None = None,
+    rotate: SecretRotate | None = None,
+    owner: str | None = None,
+):
+    """Update an existing secret."""
+    args = ['secret-set']
+    if label is not None:
+        args.extend(['--label', label])
+    if description is not None:
+        args.extend(['--description', description])
+    if expire is not None:
+        args.extend(['--expire', expire.isoformat()])
+    if rotate is not None:
+        args.extend(['--rotate', rotate.value])
+    if owner is not None:
+        args.extend(['--owner', owner])
+    args.append(id)
+
+    # Always use the --file argument to provide the content to avoid secret data
+    # ending up in the command line, where it may be logged and also is visible
+    # via /proc.
+    with tempfile.TemporaryDirectory() as tmp:
+        for k, v in (content or {}).items():
+            with open(f'{tmp}/{k}', mode='w', encoding='utf-8') as f:
+                f.write(v)
+            args.append(f'{k}#file={tmp}/{k}')
+        _run(*args)
+
+
+def state_delete(key: str):
+    """Delete server-side-state key value pairs."""
+    _run('state-delete', key)
+
+
+@overload
+def state_get(key: str) -> str: ...
+@overload
+def state_get(key: None) -> dict[str, str]: ...
+def state_get(key: str | None) -> dict[str, str] | str:
+    """Get server-side-state value."""
+    args = ['state-get']
+    if key is not None:
+        args.append(key)
+    result = _run(*args, return_output=True, use_json=True)
+    return cast('dict[str, str]', result) if key is None else cast('str', result)
+
+
+def state_set(data: Mapping[str, str], file: pathlib.Path | None = None):
```

Comment:
Similar here, do we need the `--file` argument? Can we just drop it and be opinionated that they have to pass in a dict?

Reply:
Added this one to our 1-1 list.